staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_whitebox_test.go

```diff
@@ -1670,3 +1691,515 @@ func TestCacheWatcherDrainingRequestedButNotDrained(t *testing.T) {
+				verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(104)}})
+				verifyNoEvents(t, w)
+			},
+		},
+		{
+			name: "sendInitialEvents=false, RV=0",
+			run: func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string)) {
+				// add some initial data
+				err := cacher.watchCache.Add(makePod(101))
+				require.NoError(t, err, "failed to add a pod: %v")
+				err = cacher.watchCache.Add(makePod(102))
+				require.NoError(t, err, "failed to add a pod: %v")
+				// set storage's RV to some future value, should not be used by this scenario
+				setStorageListMetaRV("105")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{ResourceVersion: "0", Predicate: storage.Everything, SendInitialEvents: func() *bool { b := false; return &b }()})
+				require.NoError(t, err, "failed to create watch: %v")
+				defer w.Stop()
+				verifyNoEvents(t, w)
+				// add a pod that is greater than the storage's RV when the watch was started
+				err = cacher.watchCache.Add(makePod(103))
+				require.NoError(t, err, "failed to add a pod: %v")
+				verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(103)}})
+				verifyNoEvents(t, w)
+			},
+		},
+		{
+			name: "legacy, RV=0",
+			run: func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string)) {
+				// add some initial data
+				err := cacher.watchCache.Add(makePod(101))
+				require.NoError(t, err, "failed to add a pod: %v")
+				err = cacher.watchCache.Add(makePod(102))
+				require.NoError(t, err, "failed to add a pod: %v")
+				// set storage's RV to some future value, should not be used by this scenario
+				setStorageListMetaRV("105")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{ResourceVersion: "0", Predicate: storage.Everything})
+				require.NoError(t, err, "failed to create watch: %v")
+				defer w.Stop()
+				// make sure get initial events
+				verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(101)}, {Type: watch.Added, Object: makePod(102)}})
+				verifyNoEvents(t, w)
+			},
+		},
+		{
+			name: "legacy, RV=unset",
+			run: func(t *testing.T, cacher *Cacher, setStorageListMetaRV func(rv string)) {
+				// add some initial data
+				err := cacher.watchCache.Add(makePod(101))
+				require.NoError(t, err, "failed to add a pod: %v")
+				err = cacher.watchCache.Add(makePod(102))
+				require.NoError(t, err, "failed to add a pod: %v")
+				// set storage's RV to some future value, should not be used by this scenario
+				setStorageListMetaRV("105")
+				w, err := cacher.Watch(context.Background(), "pods/ns", storage.ListOptions{Predicate: storage.Everything})
+				require.NoError(t, err, "failed to create watch: %v")
+				defer w.Stop()
+				// make sure get initial events
+				verifyEvents(t, w, []watch.Event{{Type: watch.Added, Object: makePod(101)}, {Type: watch.Added, Object: makePod(102)}})
+				verifyNoEvents(t, w)
+			},
+		},
+	}
```

Comment:
I would really really like to avoid huge code-duplication here. These tests are almost the same, so I think we should switch to sth:

```
scenarios := []struct {
  name string
  eventsFirst     []watch.Event
  eventsSecond []watch.Event     // better names needed
}{
  {
     name: "allowWatchBookmarks=true, sendInitialEvents=true, RV=unset",
     events: ...
  }
  ...
}


for _, scenario := range scenarios {
  t.Run(....
    // roughly what you have now

   err := cacher.watchCache.Add(makePod(101))
   require.NoError(t, err, "failed to add a pod: %v")
   err = cacher.watchCache.Add(makePod(102))
   require.NoError(t, err, "failed to add a pod: %v")

   // set storage's RV to some future value
   storageListMetaResourceVersion = "105"
   opts := storage.ListOptions{ ....}
   w, err := cacher.Watch(context.Background(), "pods/ns", opts)
   require.NoError(t, err, "failed to create watch: %v")
   defer w.Stop()

   verifyEvents(...)
   verifyNoEvents(...)

   // add a pod that is greater than the storage's RV when the watch was started
  err = cacher.watchCache.Add(makePod(110))
  require.NoError(t, err, "failed to add a pod: %v")
  verifyEvents(...)
   verifyNoEvents()
```

It will require minor changes to some usecases, but it will be waaay easier to follow and maintain.